mlrun/projects/pipelines.py

```diff
@@ -647,6 +646,9 @@ def run(
         # When using KFP, it would do this replacement. When running locally, we need to take care of it.
         if artifact_path:
             artifact_path = artifact_path.replace("{{workflow.uid}}", workflow_id)
+        if source and not workflow_spec.run_local:
+            logger.info(f"setting project source: {source}")
+            project.set_source(source)
```

Comment:
If you need to do it in every `run` method it should be in a separate function so when a bug is seen or a change is needed, you won't need to remember each place you added these lines.

```python
def set_source_if_local(project: MlrunProject, workflow_spec: WorkflowSpec, source: str):
    if source and not workflow_spec.run_local:
        logger.info(f"setting project source: {source}")
        project.set_source(source)
```